pkg/frontend/querymiddleware/split_by_instant_interval_test.go

```diff
@@ -0,0 +1,239 @@
+			query:                `min by(group_1) (rate(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"sum(rate)": {
+			query:                `sum(rate(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"sum(rate) grouping 'by'": {
+			query:                `sum by(group_1) (rate(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"topk(rate)": {
+			query:                `topk(2, rate(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"topk(sum(rate))": {
+			query:                `topk(2, sum(rate(metric_counter[3m])))`,
+			expectedSplitQueries: 3,
+		},
+		// Binary operations
+		"rate / rate": {
+			query:                `rate(metric_counter[3m]) / rate(metric_counter[6m])`,
+			expectedSplitQueries: 3,
+		},
+		"rate / 10": {
+			query:                `rate(metric_counter[3m]) / 10`,
+			expectedSplitQueries: 3,
+		},
+		"10 / rate": {
+			query:                `10 / rate(metric_counter[3m])`,
+			expectedSplitQueries: 3,
+		},
+		"sum(sum_over_time + count_over_time)": {
+			query:                `sum(sum_over_time(metric_counter[3m]) + count_over_time(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"(avg_over_time)": {
+			query:                `(avg_over_time(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"sum(avg_over_time)": {
+			query:                `sum(avg_over_time(metric_counter[3m]))`,
+			expectedSplitQueries: 3,
+		},
+		"sum(max(rate))": {
+			query:                `sum(max(rate(metric_counter[3m])))`,
+			expectedSplitQueries: 3,
+		},
+	}
+
+	series := make([]*promql.StorageSeries, 0, numSeries+(numHistograms*len(histogramBuckets)))
+	seriesID := 0
+
+	// Add counter series.
+	for i := 0; i < numSeries; i++ {
+		gen := factor(float64(i) * 0.1)
+		if i >= numSeries-numStaleSeries {
+			// Wrap the generator to inject the staleness marker between minute 10 and 20.
+			gen = stale(start.Add(10*time.Minute), start.Add(20*time.Minute), gen)
+		}
+
+		series = append(series, newSeries(newTestCounterLabels(seriesID), start.Add(-lookbackDelta), end, step, gen))
+		seriesID++
+	}
+
+	// Add a special series whose data points end earlier than the end of the queried time range
+	// and has NO stale marker.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(-lookbackDelta), end.Add(-5*time.Minute), step, factor(2)))
+	seriesID++
+
+	// Add a special series whose data points end earlier than the end of the queried time range
+	// and HAS a stale marker at the end.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(-lookbackDelta), end.Add(-5*time.Minute), step, stale(end.Add(-6*time.Minute), end.Add(-4*time.Minute), factor(2))))
+	seriesID++
+
+	// Add a special series whose data points start later than the start of the queried time range.
+	series = append(series, newSeries(newTestCounterLabels(seriesID),
+		start.Add(5*time.Minute), end, step, factor(2)))
+	seriesID++
+
+	// Add histogram series.
```

Comment:
We have no test query on histograms. Test queries you can add:

```
		"histogram_quantile() grouping only 'by' le": {
			query: `histogram_quantile(0.5, sum by(le) (rate(metric_histogram_bucket[3m])))`,
		},
		"histogram_quantile() grouping 'by'": {
			query: `histogram_quantile(0.5, sum by(group_1, le) (rate(metric_histogram_bucket[3m])))`,
		},
		"histogram_quantile() grouping 'without'": {
			query: `histogram_quantile(0.5, sum without(group_1, group_2, unique) (rate(metric_histogram_bucket[3m])))`,
		},
		"histogram_quantile() with no effective grouping because all groups have 1 series": {
			query: `histogram_quantile(0.5, sum by(unique, le) (rate(metric_histogram_bucket{group_1="0"}[3m])))`,
		},
```

Reply:
Done in commit https://github.com/grafana/mimir/pull/2469/commits/5d7fa686376da94a133465995e256d07263a486c